Adding keystone IdP

sseago opened the pull request:
This now works with Keystone v3 both with and without SSL.

Docs issue: https://github.com/openshift/openshift-docs/issues/1073

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -159,6 +162,22 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+		validationResults.AddErrors(fielderrors.NewFieldRequired("provider.url"))
+		return validationResults
+	}
+	//FIXME: we should probably allow http -- do we ignore, or do we warn here?
```
warn

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,66 @@
+
+	"github.com/golang/glog"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack"
```
need to add the godeps for this... how big are those?

deads2k replied:
> need to add the godeps for this... how big are those?

When you add them, see the commit format.  We keep the godeps in different commits starting with `bump(dependency name): commit_id`

sseago replied:
Ahh ok. These are already referenced in upstream kubernetes -- I wasn't sure if direct vs. indirect deps made a difference.

sseago replied:
@liggitt @deads2k Looks like the gophercloud deps are already here as of this commit: https://github.com/openshift/origin/commit/83f9bfd2d4e5ecabfbd46c67f78edda058656051

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -159,6 +162,22 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+
+	if len(provider.RemoteConnectionInfo.URL) == 0 {
```
should be able to just call ValidateRemoteConnectionInfo, which will check the url, ca, and client cert stuff

liggitt replied:
then warn about non-https separately

sseago commented:
Updated the commit:
1) incorporated feedback
2) SSL working
3) moved the hard-coded DomainName into a config param.

Unit testing is still missing.

sseago commented:
Added panic recovery

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -134,6 +134,10 @@ func ValidateIdentityProvider(identityProvider api.IdentityProvider) ValidationR
 			validationResults.Append(ValidateLDAPIdentityProvider(provider))
 
+		case (*api.KeystonePasswordIdentityProvider):
+			validationResults.AddErrors(ValidateRemoteConnectionInfo(provider.RemoteConnectionInfo).Prefix("provider")...)
```
move the ValidateRemoteConnectionInfo call into ValidateKeystoneIdentityProvider

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -160,6 +164,18 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+func ValidateKeystoneIdentityProvider(provider *api.KeystonePasswordIdentityProvider, identityProvider api.IdentityProvider) ValidationResults {
+	validationResults := ValidationResults{}
+
+	if !strings.HasPrefix(provider.RemoteConnectionInfo.URL, "https") {
```
parse the url and check the scheme, rather than prefix

sseago replied:
Fixing this

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -160,6 +164,18 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+	validationResults := ValidationResults{}
+
+	if !strings.HasPrefix(provider.RemoteConnectionInfo.URL, "https") {
+		validationResults.AddWarnings(fielderrors.NewFieldInvalid("provider.url", provider.RemoteConnectionInfo.URL, "Auth URL should be secure and start with https"))
```
`NewFieldInvalid("url", ...)`, and let the caller prefix with "provider"

sseago replied:
ok

liggitt commented on pkg/cmd/server/origin/auth.go:
```diff
@@ -520,6 +522,21 @@ func (c *AuthConfig) getPasswordAuthenticator(identityProvider configapi.Identit
+			return nil, fmt.Errorf("Error building KeystonePasswordIdentityProvider client: %v", err)
+		}
+
+		if !strings.HasPrefix(connectionInfo.URL, "https") {
```
do we really want to log both here and in validation?

sseago replied:
Probably not. I'll pull this bit out.

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -160,6 +164,18 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+
+	if !strings.HasPrefix(provider.RemoteConnectionInfo.URL, "https") {
+		validationResults.AddWarnings(fielderrors.NewFieldInvalid("provider.url", provider.RemoteConnectionInfo.URL, "Auth URL should be secure and start with https"))
+		return validationResults
```
any validation of DomainName needed?

liggitt replied:
actually, are we requiring the v3 API? If so, is domainName required, and should we be using `AuthenticateV3()`?

sseago replied:
I don't see anything obvious listed in the keystone API docs around domain name formatting, but we should make sure it's present and probably validate the URL against v3 (and call AuthenticateV3).

sseago replied:
Making the explicit AuthenticateV3 call is the way to go. This forces v3 even if the config URL omits the version.

sseago commented:
I just updated the PR incorporating feedback and adding a working unit test against a mocked test keystone server.

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,84 @@
+
+	// Calling NewClient/Authenticate manually rather than simply calling AuthenticatedClient
+	// in order to pass in a transport object that supports SSL
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
```
check err and return before using `client`... it might be nil

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,84 @@
+	// in order to pass in a transport object that supports SSL
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
+	client.HTTPClient = *a.client
+	if err == nil {
```
no need to check err here if you check above

sseago replied:
So I think all I need to do here is move the assignment of client.HTTPClient into the "if err == nil" block. If there's no error, we set HTTPClient and authenticate in the "err isn't nil" block. If there is an error, that's handled in the block below.

liggitt replied:
convention is to return immediately on errors

```
foo, err := op()
if err != nil {
  return nil, err
}

bar, err := op()
if err != nil {
  return nil, err
}
```

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,84 @@
+				return nil, false, nil
+			}
+		}
+		if err == tokens.ErrMissingPassword {
```
is an empty password ever valid? if not, can we return immediately on empty passwords?

sseago replied:
It's not valid. The gophercloud library errors out with ErrMissingPassword without even making the call to keystone, but yeah if we do the same check ourselves it might make more sense than passing it on and checking for the error return.

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,84 @@
+		if err == tokens.ErrMissingPassword {
+			return nil, false, nil
+		}
+		return nil, false, errors.New("Failed to authenticate")
```
return the original error so we don't lose details? maybe add logging as well, not sure what level it should be at

sseago replied:
I already have a logging statement above (first line in the block, above the 401 check). I can return the original error -- it gets swallowed by the UI code to display a generic "something went wrong" sort of error to the user anyway. Actually, come to think of it we probably don't need to log the error if it's a 401 response -- that's just a failed login, and if we get any other error here (other than the missing password one that we should handle differently), perhaps it should be a higher level log, since any error other than a 401 or a missing password probably indicates a config problem or a problem with keystone. Should it be glog.Errorf? (or Warningf)?

liggitt replied:
I need to sweep how logging works for auth in general... how individual authenticators log, and how the server handles returned errors. For now, log at high verbosity on login failure, and log an error if you get to this line. I'll revisit.

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go:
```diff
@@ -0,0 +1,101 @@
+	"testing"
+	"io/ioutil"
+	
+//	"github.com/rackspace/gophercloud"
```
clean up comments

sseago replied:
Oops. I missed that one.

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -162,6 +166,25 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+	validationResults.AddErrors(ValidateRemoteConnectionInfo(provider.RemoteConnectionInfo)...)
+
+	providerURL, err := url.Parse(provider.RemoteConnectionInfo.URL)
+	if err != nil {
```
ValidateRemoteConnectionInfo already adds an error if the url doesn't parse... you can just check the scheme if `err == nil`

liggitt commented:
[test]

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,84 @@
+	}
+
+	if err != nil {
+		glog.Infof("Failed: Starting openstack authenticate client: %v", err)
```
higher verbosity, probably V(4)

liggitt commented:
thanks, just a few final things and a clean test run

liggitt commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -544,6 +544,15 @@ type LDAPAttributeMapping struct {
 	Email []string `json:"email"`
 }
 
+// FIXME: add this to types_test.go
```
add this

sseago commented:
Next round of cleanup pushed to PR...

liggitt commented on pkg/auth/authenticator/password/keystonepassword/keystonepassword.go:
```diff
@@ -0,0 +1,86 @@
+
+	identity := authapi.NewDefaultUserIdentityInfo(a.providerName, username)
+	user, err := a.identityMapper.UserFor(identity)
+	glog.V(4).Infof("Got userIdentityMapping: %#v", user)
```
move this after the `err != nil` check

liggitt commented on pkg/cmd/server/api/v1/types.go:
```diff
@@ -544,6 +544,14 @@ type LDAPAttributeMapping struct {
 
+type KeystonePasswordIdentityProvider struct {
+	v1.TypeMeta `json:",inline"`
+	// RemoteConnectionInfo contains information about how to connect to the external basic auth server
```
s/external basic auth server/keystone server/

liggitt commented:
godoc nit, LGTM

liggitt commented:
tests failed with:
```
hack/verify-gofmt.sh
!!! gofmt needs to be run on the following files: 
./pkg/auth/authenticator/password/keystonepassword/keystonepassword_test.go
```

sseago commented:
verify-gofmt succeeds now. I'd run it on the other new files but forgot to run it on the test file.

liggitt commented:
[test]

liggitt commented on pkg/cmd/server/api/validation/oauth.go:
```diff
@@ -162,6 +166,24 @@ func ValidateLDAPIdentityProvider(provider *api.LDAPPasswordIdentityProvider) Va
+		if providerURL.Scheme != "https" {
+			validationResults.AddWarnings(fielderrors.NewFieldInvalid("url", provider.RemoteConnectionInfo.URL, "Auth URL should be secure and start with https"))
+		}
+	}
```
I forgot, `ValidateSecureURL` already exists... I think this whole bit can simplify to 
```
validationResults.AddWarnings(ValidateSecureURL("url", provider.RemoteConnectionInfo.URL)
```

sseago replied:
So I looked into this. What was needed was similar to what you have there:
```
       _, urlErrs := ValidateSecureURL(provider.RemoteConnectionInfo.URL, "url")
       validationResults.AddWarnings(urlErrs...)
```
The only issue is while it acts as a warning (and not a server-stopping error), the log message implies a real error:
```
W1014 15:16:28.928500   28181 start_master.go:278] oauthConfig.identityProvider[0].provider.url: invalid value 'http://192.168.0.198:5000/v3', Details: must use https scheme
```
With the custom validation we avoid the inaccurate message. I could refactor ValidateSecureURL to take a 'warning' vs. 'error' boolean arg, but that might be more confusing than just creating a WarnOnInsecureURL method, but with only one caller of it, I don't know if that's any better than what I already have above.

sseago commented:
@liggitt  What's left to do here. Jenkins is succeeding but Travis failed. Is this a known issue with the travis results, or does this represent an actual problem with this PR that needs fixing?

liggitt commented:
You're all set, thanks. I'll work on getting this merged tomorrow.

sseago commented:
Excellent. Thanks for the update.

liggitt commented:
[test]

liggitt commented:
[test]

liggitt commented:
[merge]
